### PR TITLE
Simplify and improve performance of NameGenerator.EnsureUniqueness methods

### PR DESCRIPTION
**Pull request by ToddGrun**
I noticed this method's call to FillCollisionIndices in prism CPU.

The primary EnsureUniqueness method can be significantly simplified and performance improved at the expense of not sharing code with the other methods.

The other methods are also simplified and should have improved performance by removing the O(n^2) nature of the calls to FillCollisionIndices.

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/NameGenerator.cs [hunk 1]
```diff
@@ -21,24 +21,30 @@ internal static class NameGenerator
...
 
-            names.Add(baseName);
-            isFixed.Add(false);
+            if (!isCaseSensitive)
+            {
+                // The pooled collections don't allow non-default comparers (and they can't be set on
+                //   dictionary/hashset after creation). As this scenario is infrequent, we'll just
+                //   create allocate new objects for these. The pooled object will still be freed
+                //   via the PooledDisposer.
+                nameSet = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/NameGenerator.cs:33]: you can make a dedicated pool for the ignorecase case.  i'd prefer that.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/NameGenerator.cs:33]: also fine to have our own dedicated string pool for the case sensitive case.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/NameGenerator.cs [hunk 2]
```diff
@@ -91,71 +97,56 @@ internal static class NameGenerator
...
+            if (!isCaseSensitive)
             {
-                var name = names[i];
-                FillCollisionIndices(names, name, isCaseSensitive, collisionIndices);
+                // The pooled collections don't allow non-default comparers (and they can't be set on
+                //   dictionary/hashset after creation). As this scenario is infrequent, we'll just
+                //   create allocate new objects for these. The pooled objects will still be freed
+                //   via the PooledDisposers.
+                collisionMap = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
+                usedNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/NameGenerator.cs:113]: same as above please.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/NameGenerator.cs (hunk 1 above)
Comment (ToddGrun) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/NameGenerator.cs]: Let me know if you'd rather me keep the changes restricted to adding pooling just in this class

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledStringDictionary.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+internal sealed partial class PooledStringDictionary<V>
+{
+    private static ObjectPool<PooledDictionary<string, V>>? s_poolCaseInsensitiveInstance;
+
+    public static PooledDisposer<PooledDictionary<string, V>> GetInstance(bool caseSensitive, out PooledDictionary<string, V> instance)
+    {
+        if (caseSensitive)
+            return PooledDictionary<string, V>.GetInstance(out instance);
+
+        s_poolCaseInsensitiveInstance ??= PooledDictionary<string, V>.CreatePool(StringComparer.OrdinalIgnoreCase);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledStringDictionary.cs:19]: just make it a static readonly and always initialize.  it's not going to matter.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledStringHashSet.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+internal class PooledStringHashSet
+{
+    private static ObjectPool<PooledHashSet<string>>? s_poolCaseInsensitiveInstance;
+
+    public static PooledDisposer<PooledHashSet<string>> GetInstance(bool caseSensitive, out PooledHashSet<string> instance)
+    {
+        if (caseSensitive)
+            return PooledHashSet<string>.GetInstance(out instance);
+
+        s_poolCaseInsensitiveInstance ??= PooledHashSet<string>.CreatePool(StringComparer.OrdinalIgnoreCase);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledStringHashSet.cs:19]: same

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledStringDictionary.cs (hunk 3 above)
Comment (ToddGrun) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledStringDictionary.cs:19]: Per our conversation, I'm going to keep it as such as if it were the other way, a pool would be created for every value type of a string dictionary and that idea bothered me.

Comment (alrz) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledStringDictionary.cs:19]: I think you could do something like this if you prefer. https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/SpecializedSymbolCollections.cs,e96e13892c141f98,references

Comment (ToddGrun) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledStringDictionary.cs:19]: @alrz -- I'm not clear on how that would help me avoid allocating this static pool for every type V that this is called for. I really only want to create the case insensitive pool for V if this method is called indicating a desire to use a case insensitive comparer.

Comment (alrz) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledStringDictionary.cs:19]: The initializer won't run if you don't reach the field access. Perhaps something like: ([sharplab](https://sharplab.io/#v2:CYLg1APgAgDABFAjAOgCIEsCGBzAdgewGcAXdAY0IG4BYAKDvV2IFMAnXTAGwUQDYEATHADKAB2ZksndAC9mwYcVaNsAYXydOE0vlyE6AbzpwTCAMw9+ABXwb5ACUyEAFsObEAPEhgA+OAHF3GzsFJRVHFzdiAEk9YkxcMmYACgAjW24yJ2Y3PXRSADdmAEpjUyNaUyq4dAAzOGSswhzmPMKShAB2OGCtUOVcbAjXdy9EX2RAmLiEpOTimkrqkwLMVhqZxOY4AF4ejPlFAaGnEeJkQgB9UQzYklnmZABBTXwslnnF5ZNUZlSAV2wz0IzVYxGSjHuW2Q6n+TF2exgCzKyyg3Uh8S2X1MAF86Ci4KJlKsWJZBPsQkdwqcogSKt8Nix2FwyaxmJhgLpOABPOAAeVSACttL0PL0HDTRt4fH4rjcNHdMUldhS+sMomMJqo2ZgWL1klTBuoALaiNZsZB81jARhcaJ4fBs1TZZFLEx4+hu8xk8XADBkHQcVjczUAGjgADU/FNfYbsP7A2tuYqHh4o2kMnAmi02ugiqUvfTvnUGtncoR8nmOmjVfIE+hdEmw5GfJN3Cmtp8CVVVusMQ8VbGwoN643g2nW3Lbpsks9Xu8Uq6Gb8AUCniC2ODIQ8Yfg4cQEXAkdjUeiZ8wT3APQSiXnddskPwoEIh8dR0GQ1G6d3TIwmRxuEfOAdU5XAeX5IURQyMUDj9chE3HbxwyjWVrmnKFlT2X13ybJCWxhHU9QyA1hzUfBTXNVhLWtW1OHtAgnRdS8PRxIA==))

Comment (ToddGrun) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledStringDictionary.cs]: @alrz -- These were really great suggestions and led me to look into this a bit more at whcih point I found that Roslyn already has existing pools for these structures. I'm still learning to *extensively* look around Roslyn for existing solutions before implementing something myself.

**Pull request merged**